msprime/species_tree.py

```diff
@@ -0,0 +1,834 @@
+
+    def get_edges(self):
+        return self.edges
+
+    def get_number_of_edges(self):
+        return len(self.edges)
+
+    def get_number_of_extant_edges(self):
+        number_of_extant_edges = 0
+        for edge in self.edges:
+            if edge.get_is_extant():
+                number_of_extant_edges += 1
+        return number_of_extant_edges
+
+    def set_extant_progeny_ids(self):
+        for edge in self.get_edges():
+            if edge.get_is_extant():
+                species_id = edge.get_node_ids()[1]
+                this_edge = edge
+                species_id_added_to_root_edge = False
+                while species_id_added_to_root_edge is False:
+                    this_edge.add_extant_progeny_id(species_id)
+                    if this_edge.get_is_root_edge():
+                        species_id_added_to_root_edge = True
+                    else:
+                        for other_edge in self.get_edges():
+                            if (other_edge.get_node_ids()[1] ==
+                                    this_edge.get_node_ids()[0]):
+                                parent_edge = other_edge
+                                break
+                    this_edge = parent_edge
+
+    def set_times(self):
+        # Get the durations between root and extant edges.
+        total_edge_lengths = []
+        for edge in self.get_edges():
+            if edge.get_is_extant():
+                total_edge_length = edge.get_length()
+                if edge.get_is_root_edge():
+                    total_edge_lengths.append(total_edge_length)
+                else:
+                    root_edge_found = False
+                    this_edge = edge
+                    while root_edge_found is False:
+                        for other_edge in self.get_edges():
+                            if (other_edge.get_node_ids()[1] ==
+                                    this_edge.get_node_ids()[0]):
+                                parent_edge = other_edge
+                                break
+                        err = 'The parent edge for edge '
+                        err += '{} could '.format(this_edge.get_id())
+                        err += 'not be found'
+                        assert parent_edge, err
+                        total_edge_length += parent_edge.get_length()
+                        if parent_edge.get_is_root_edge():
+                            root_edge_found = True
+                            total_edge_lengths.append(total_edge_length)
+                        else:
+                            this_edge = parent_edge
+        # Make sure that the tree is at least roughly ultrametric.
+        # A maximum difference of 1% between the shortest and longest
+        # root-to-tip distances is allowed.
+        max_total_edge_length = max(total_edge_lengths)
+        if (max_total_edge_length - min(total_edge_lengths) >
+                0.01 * max_total_edge_length):
+            raise NotImplementedError(
```

Comment:
This is a ValueError --- NotImplementedError is used when a subclass hasn't implemented an abstract method from a superclass

Reply:
Fixed.